core/server/data/migrations/utils.js

```diff
@@ -0,0 +1,244 @@
+            }).first();
+
+            if (existingRelation) {
+                logging.warn(`Adding perm(${config.permission}) to role(${config.role}) - already exists`);
+                return;
+            }
+
+            await connection('permissions_roles').insert({
+                id: ObjectId.generate(),
+                permission_id: permission.id,
+                role_id: role.id
+            });
+        },
+        async function down(connection) {
+            const permission = await connection('permissions').where({
+                name: config.permission
+            }).first();
+
+            if (!permission) {
+                throw new Error(
+                    `Cannot remove perm(${config.permission}) from role(${config.role}) - perm does not exist`
+                );
+            }
+
+            const role = await connection('roles').where({
+                name: config.role
+            }).first();
+
+            if (!role) {
+                throw new Error(
+                    `Cannot remove perm(${config.permission}) from role(${config.role}) - role does not exist`
+                );
+            }
+
+            const existingRelation = await connection('permissions_roles').where({
+                permission_id: permission.id,
+                role_id: role.id
+            }).first();
+
+            if (!existingRelation) {
+                logging.warn(`Removing perm(${config.permission}) to role(${config.role}) - already removed`);
+                return;
+            }
+
+            await connection('permissions_roles').where({
+                permission_id: permission.id,
+                role_id: role.id
+            }).del();
+        }
+    );
+}
+
+/**
+ * Creates a migration which will add a permission to the database, and then link it to roles
+ *
+ * @param {Object} config
+ * @param {string} config.name - The name of the permission
+ * @param {string} config.action - The action_type of the permission
+ * @param {string} config.object - The object_type of the permission
+ *
+ * @param {string[]} roles - A list of role names
+ *
+ * @returns {Migration}
+ */
+function addPermissionWithRoles(config, roles) {
+    return combineTransactionalMigrations(
+        addPermission(config),
+        ...roles.map((role => addPermissionToRole({permission: config.name, role})))
+    );
+}
+
+/**
+ * @param {(connection: import('knex')) => Promise<void>} up
+ * @param {(connection: import('knex')) => Promise<void>} down
+ *
+ * @returns {Migration}
+ */
+function createTransactionalMigration(up, down) {
+    return {
+        config: {
+            transaction: true
+        },
+        async up(config) {
+            await up(config.transacting);
+        },
+        async down(config) {
+            await down(config.transacting);
+        }
+    };
+}
+
+/**
+ * @param {Migration[]} migrations
+ *
+ * @returns {Migration}
+ */
+function combineTransactionalMigrations(...migrations) {
```

Comment:
The idea here is to encourage building smaller migrations that we can combine together to make more complex ones, hopefully this makes utils easier to review & test!